src/net.md

```diff
@@ -1019,6 +1020,151 @@ fn run() -> Result<()> {
+    content_length: u64,
+    range_ends: u64,
+    ranges: Vec<u64>,
+}
+
+impl PartialRange {
+    pub fn new(content_range: &ContentRangeSpec) -> PartialRange {
+        let (content_length, range_starts, range_ends) = match *content_range {
+            ContentRangeSpec::Bytes {
+                range: Some(range),
+                instance_length: Some(content_length),
+            } => (content_length, range.0, range.1),
+            _ => (0, 0, 0),
+        };
+
+        let ranges: Vec<u64> = (range_starts..((range_ends - range_starts) / BUFFER_SIZE + 1))
+            .map(|c| c * BUFFER_SIZE)
+            .collect();
+
+        PartialRange {
+            content_length,
+            range_ends,
+            ranges,
+        }
+    }
+
+    pub fn iter(&self) -> PartialRangeIterator {
+        PartialRangeIterator::new(self.range_ends, &self.ranges)
+    }
+}
+
+struct PartialRangeIterator<'a> {
+    ranges: &'a [u64],
+    range_ends: u64,
+    next_start: Option<u64>,
+}
+
+impl<'a> PartialRangeIterator<'a> {
+    pub fn new(range_ends: u64, ranges: &'a [u64]) -> PartialRangeIterator<'a> {
+        PartialRangeIterator {
+            ranges,
+            range_ends,
+            next_start: if ranges.is_empty() { None } else { Some(0) },
+        }
+    }
+}
+
+impl<'a> Iterator for PartialRangeIterator<'a> {
+    type Item = Range;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.next_start.is_none() {
+            None
+        } else {
+            let next_start = self.next_start.unwrap();
+            self.next_start = if next_start == (self.ranges.len() - 1) as u64 {
+                None
+            } else {
+                Some(next_start + 1)
+            };
+
+            if self.next_start.is_none() {
+                Some(Range::bytes(next_start * BUFFER_SIZE, self.range_ends))
+            } else {
+                Some(Range::bytes(
+                    next_start * BUFFER_SIZE,
+                    (next_start * BUFFER_SIZE) + BUFFER_SIZE - 1,
+                ))
+            }
+        }
+    }
+}
+
+fn run() -> Result<()> {
+    let client = reqwest::Client::new();
+
+    let url = "https://httpbin.org/range/1024?duration=2";
+    let response = client.head(url).send()?;
+
+    let range = response.headers().get::<ContentRange>().ok_or(
+        "response doesn't include the expected ranges",
+    )?;
+
+    let mut output_file = File::create("download.bin")?;
+    let ranges = PartialRange::new(range);
+
+    println!("starting download...");
+    for range in ranges.iter() {
+        println!("range {}", range);
+
+        let mut response = client.get(url).header(range).send()?;
+
+        if response.status() != reqwest::StatusCode::PartialContent &&
+            response.status() != reqwest::StatusCode::Ok
+        {
+            bail!("Unexpected server response: {}", response.status())
+        }
+
+        let mut content: Vec<u8> = Vec::with_capacity(BUFFER_SIZE as usize);
```

Comment:
no need to allocate the buffer on each iteration. Lets allocate it before the loop and use `Vec::clear` after each iteration